File: typescript/relay/langchain.ts
/**
 * Relay + LangChain.js / LangGraph.js (`surface: "langchain"` | `"langgraph"`).
 *
 * Peers: npm i langchain @langchain/core
 * Env:   EXEMPLAR_API_KEY (+ model keys for a live createAgent run)
 *
 *   npm run build && npx tsx examples/relay/langchain.ts
 */
import { harness, newSessionId, banner, SOURCE_APP } from "../_helpers.ts";

async function main(): Promise<void> {
  const h = harness("langchain-relay-bot");
  const sessionId = newSessionId("relay-langchain");
  banner("Relay + LangChain.js", sessionId);

  const relay = h.relay({
    surface: "langchain",
    sourceApp: SOURCE_APP,
  });

  const mwConfig = relay.langchainMiddleware(sessionId);
  console.log("middleware name:", mwConfig.name);

  let handlerRan = false;
  const result = await mwConfig.wrapToolCall(
    {
      toolCall: {
        name: "shell",
        args: { command: "cat /etc/passwd" },
        id: "call_example",
      },
    },
    async () => {
      handlerRan = true;
      return { ok: true };
    },
  );
  console.log("wrapToolCall result:", result);
  console.log(
    handlerRan
      ? "handler ran (policy allowed or fail-open on transport error)"
      : "handler skipped (Relay denied / asked)",
  );

  // Second call: benign tool — observe path when allowed.
  handlerRan = false;
  const allowed = await mwConfig.wrapToolCall(
    {
      toolCall: {
        name: "lookup_order",
        args: { orderId: "ORD-1" },
        id: "call_ok",
      },
    },
    async () => {
      handlerRan = true;
      return { orderId: "ORD-1", status: "shipped" };
    },
  );
  console.log("lookup_order:", { allowed, handlerRan });

  console.log(`
Wire into LangChain createAgent:

  import { createAgent, createMiddleware } from "langchain";

  const middleware = createMiddleware(relay.langchainMiddleware(sessionId));
  // langgraph surface: relay.langgraphMiddleware(sessionId) — same shape

  const agent = createAgent({
    model: yourModel,
    tools: [/* … */],
    middleware: [middleware],
  });

  await agent.invoke({ messages: [{ role: "user", content: "Check order ORD-1" }] });
`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: python/relay/langchain.py
"""Live Relay + LangChain create_agent wrap_tool_call middleware.

Also see langchain_middleware.py for a lighter smoke sample.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from common import SOURCE_APP, harness, session_id  # noqa: E402


class _Req:
    def __init__(self, name: str, args: dict, call_id: str = "c1") -> None:
        self.tool_call = {"name": name, "args": args, "id": call_id}
        self.tool = None


def main() -> None:
    h = harness("relay-langchain-sample")
    sid = session_id("relay-langchain")
    relay = h.relay(surface="langchain", source_app=SOURCE_APP)
    mw = relay.langchain_middleware(session_id=sid)
    wrap = getattr(mw, "wrap_tool_call", None) or mw

    allow = wrap(_Req("get_weather", {"city": "SF"}), lambda _r: "Sunny in SF")
    deny = wrap(
        _Req("shell", {"command": "rm -rf /tmp/x"}),
        lambda _r: "should-not-run",
    )
    allow_text = getattr(allow, "content", allow)
    deny_text = getattr(deny, "content", deny)

    print(
        json.dumps(
            {
                "example": "relay_langchain",
                "sessionId": sid,
                "middlewareType": type(mw).__name__,
                "allow": str(allow_text),
                "deny": str(deny_text)[:200],
                "wire": "create_agent(..., middleware=[relay.langchain_middleware(...)])",
            },
            indent=2,
        )
    )


if __name__ == "__main__":
    main()
